[MIG][REF] delivery_carrier_label_gls

len-foss opened the pull request:
It's actually essentially rewritten from scratch to use the REST API.
It's available in multiple countries, including Belux, Germany, Austria, excluding France.
You need to contact GLS to check if it is available in your country, as they seem unable to provide the complete list. 

Caution: 
The API does not offer any idempotent way to make the calls, so any lost call to "createParcel" will be invoiced, even if the parcel is never shipped (wew lad!). Therefore, a call to cancelShipment should be done in case there is a rollback.
However this was only done in https://github.com/odoo/odoo/pull/54321, so you need to pull it in your sources if you want that safety net.

Commit:
[MIG][REF] delivery_carrier_label_gls: using the international REST API

Note that the international API is not available everywhere,
notably not in France, but available in Belgium, Luxembourg, Germany, Austria.
Ask your local GLS for the availability of the web service.

pedrobaeza commented:
@chienandalu is this webservice the same as the Spanish one?

len-foss commented:
@pedrobaeza AFAIK this is supposed to be the international API, except for France (they use a Minitel-based protocol probably). Is there already an implementation?

pedrobaeza commented:
https://github.com/OCA/l10n-spain/tree/12.0/delivery_gls_asm

len-foss commented:
Oh thanks for the pointer.
It's definitely not the same API. I'll have to check in which countries this one can really be used.

pedrobaeza commented:
The one in Spanish localization is AFAIK inherited from the old carrier (ASM) that was absorbed by GLS.

Commit:
[IMP] delivery_carrier_label_gls: manage rollback on createParcel

Note: dependent on https://github.com/odoo/odoo/pull/54321/
to actually work.

lmignon commented on delivery_carrier_label_gls/tests/common.py:
```diff
@@ -0,0 +1,152 @@
+                    }
+                ],
+                u"PickupLocation": u"BE7100",
+                u"PrintData": [{u"Data": u"JVeryLongString==", u"LabelFormat": u"PDF"}],
```
@len-foss The LabelFormat should be configurable on the carrier info

lmignon replied:
@len-foss see https://github.com/acsone/delivery-carrier/pull/1

lmignon committed:
[IMP] delivery_carrier_label_gls: Allows to specify the expected format for the delivery label

Commit:
[UPD] delivery_carrier_label_gls: French translation

lmignon committed:
[IMP] delivery_carrier_label_gls: GLS parameters are required on GLS delivery carrier

Adds constraint to check valid configuration of GLS carriers.
Also check that no GLS parameters are set on non_GLS carriers,
to avoid incorrect configuration.
Remove related fields and config to simplify the code and checks.

Commit:
[FIX] delivery_carrier_label_gls: correct check that everything is packaged

Commit:
[IMP] delivery_carrier_label_gls: do not crash on light packages

The GLS API does not accept weights under 100g,
so we automatically integrate that constraint.

Commit:
[IMP][TEST] delivery_carrier_label_gls: provide a mock manager for inheriting module tests

Commit:
[IMP] delivery_carrier_label_gls: add configurable return address on GLS carriers

Commit:
[IMP] delivery_carrier_label_gls: better logging

Commit:
[IMP] delivery_carrier_label_gls: add mobile phone

Commit:
[IMP] delivery_carrier_label_gls: auto-retry without service

Commit:
[IMP] delivery_carrier_label_gls: enforce fields max length

Commit:
[IMP] delivery_carrier_label_gls: send street2 field

Commit:
[IMP] delivery_carrier_label_gls: make shipment payload inheritable

Commit:
[IMP] delivery_carrier_label_gls: normalize ZIP codes to avoid errors

GLS does not tolerate common ways to write ZIP codes.
We introduce automatic normalization for a few countries
which can be completeted as necessary.

Commit:
[FIX] delivery_carrier_label_gls: replace Monaco by France

Although GLS knows Monaco, their system accepts more options
if the same Monacan address is sent with country set as FR.
This makes absolutely no sense, of course.

rousseldenis commented:
@len-foss Could you check Travis errors (flake8 and pylint)

lmignon commented on delivery_carrier_label_gls/models/stock_picking.py:
```diff
@@ -0,0 +1,65 @@
+
+    def _check_is_everything_packaged(self):
+        for picking in self:
+            if not all(o.result_package_id for o in picking.pack_operation_product_ids):
```
@len-foss what about `pack_operation_pack_ids`?
